### PR TITLE
Cleanup HTTP virtual host generation logic

### PR DESCRIPTION
**Pull request by nmittler**
The current code is a little hard to follow and isn't well documented. This PR attempts to restructure the method slightly to more clearly separate k8s and non-k8s scenarios.

There is a small amount of duplicate code in each `if` branch for the case when `len(uniqueHostnameParts) == 2`, but it's not much and is probably worth the duplication.

These changes also avoid a couple needless calls to `strings.Split` and `strings.Join`.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 1]
```diff
@@ -527,13 +539,11 @@ func getUniqueAndSharedDNSDomain(fqdnHostname, proxyDomain string) (string, stri
 	}
 
 	if len(sharedSuffixesInReverse) == 0 {
-		return fqdnHostname, ""
+		return reverseArray(partsFQDN), nil
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:542]: very nit: more efficient to save the result of `strings.Split(fqdnHostname, ".")` on L529 and reuse here

**Review by ramaraochavali — COMMENTED** [review]
LGTM. Will leave it for @howardjohn for final review and approval

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 2]
```diff
@@ -411,49 +411,61 @@ func generateVirtualHostDomains(service *model.Service, port int, node *model.Pr
...
-			vhosts = append(vhosts, mcsHost, domainName(mcsHost, port))
-		}
-	}
+	uniqueHostname := strings.Join(uniqueHostnameParts, ".")
+
+	if strings.Contains(proxyDomain, ".svc.") {
+		// Proxy is k8s.
 
-	sharedDNSDomainParts := strings.Split(sharedDNSDomain, ".")
-	if len(strings.Split(uniqHostname, ".")) == 2 {
```
[the remark this answers is not in the document]
Comment (nmittler) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:433]: We now apply this check in each branch. That's the duplication I referred to in the description. It's a little ugly, but it's not much code and having it split out this way makes the logic easier to follow (at least in my mind).

**Comment by nmittler** [comment]
@howardjohn PTAL

**Review by howardjohn — APPROVED** [review]
